src/lib.rs

```diff
@@ -1,144 +1,181 @@
+pub use opt::Opt;
 
 /// An argument parser.
 ///
 /// See the [crate documentation](index.html) for more details.
-pub struct Options<'a, S>
+pub struct Options<'str, I>
 where
-    S: AsRef<str>,
+    I: Iterator<Item = &'str str>
 {
-    args: &'a [S],
+    /// Iterator over the arguments.
+    iter: I,
     /// State information.
-    inner: RefCell<OptionsInner<'a>>,
-}
-
-struct OptionsInner<'a> {
-    /// Index in `args`.
-    position: usize,
-    /// The state information.
-    state: State<'a>,
+    state: State<'str>
 }
 
-#[derive(Copy, Clone)]
-enum State<'a> {
+#[derive(Copy, Clone, Debug)]
+enum State<'str> {
     /// The starting state. We may not get a value because there is no
     /// previous option. We may get a positional argument or an
     /// option.
     Start,
+    /// We found a positional option and want to preserve it, since it
+    /// will no longer be returned from the iterator.
+    Positional(&'str str),
     /// We have just finished parsing an option, be it short or long,
     /// and we don't know whether the next argument is considered a
     /// value for the option or a positional argument. From here, we
     /// can get the next option, the next value, or the next
     /// positional argument.
-    EndOfOption(Opt<'a>),
+    EndOfOption(Opt<'str>),
     /// We are in the middle of a cluster of short options. From here,
     /// we can get the next short option, or we can get the value for
     /// the last short option. We may not get a positional argument.
-    ShortOptionCluster(Opt<'a>, usize),
+    ShortOptionCluster(Opt<'str>, &'str str),
     /// We just consumed a long option with a value attached with `=`,
     /// e.g. `--execute=expression`. We must get the following value.
-    LongOptionWithValue(Opt<'a>, usize),
+    LongOptionWithValue(Opt<'str>, &'str str),
+    /// We have recieved `None` from the iterator and we are refusing to advance to be polite.
+    End,
 }
 
-impl<'a> State<'a> {
-    fn last_opt(&self) -> Opt<'a> {
-        match *self {
-            State::Start => panic!("called last_opt() on State::Start"),
-            State::EndOfOption(opt) => opt,
-            State::ShortOptionCluster(opt, _) => opt,
-            State::LongOptionWithValue(opt, _) => opt,
-        }
+/// An iterator over the positional arguments of an [`Options`]. Calls
+/// to [`Iterator::next`] will forward to [`Options::arg`]. This
+/// iterator can be obtained by calling [`Options::args`].
+///
+/// # Example
+///
+/// ```
+/// # use getargs::{Opt, Options};
+/// #
+/// let args = ["--flag", "one", "two"];
+/// let mut opts = Options::new(args.into_iter());
+///
+/// assert_eq!(opts.next(), Ok(Some(Opt::Long("flag"))));
+/// assert_eq!(opts.next(), Ok(None));
+///
+/// let mut args = opts.args();
+///
+/// assert_eq!(args.next(), Some("one"));
+/// assert_eq!(args.next(), Some("two"));
+/// assert_eq!(args.next(), None);
+/// ```
+pub struct ArgIterator<'opts, 'str, I: Iterator<Item=&'str str>> {
```

Comment:
Pretty sure there should be spaces around the `=`. Run `cargo fmt` to reformat the code.

Reply:
Whoops, didn't know you used rustfmt (there's no configuration file!) That would explain the indentation